### ISSUE TITLE
Probably not  a good idea...

### ISSUE THREAD

**Issue by fulmicoton** [PSeitz/rust_string_hashmap#1]
I see you are working on trying to extrude tantivy's strange HashMap into a different project, and change the hashtable policy to robin hood hashing... Please let me now if you reach a point where it can be used by tantivy.

That's cool (as long as you don't forget to copy the license for the heap/expull stuff at one point), but the code you forked is actually "theoretically" bugged because I wrote it without caring at all about alignment.
In practise this has not been a problem so far, because x86 is fine and LLVM does not have any optimization relying on alignment there.
The way I was handling the memory arena was a bit stupid too.
I started working on fixing it in `https://github.com/tantivy-search/tantivy/tree/better-heap` so the code you forked will actually disappear very soon.

I also experimented a bit on this project [onehop-hashmap](https://github.com/fulmicoton/onehop-hashmap)... A couple of insight from my tries there.

If you go through the commit history I did try robin hood hashing too ...
https://github.com/fulmicoton/onehop-hashmap/commit/e2549eb77bd8665b1039ed7f2fd85cb7f75b37f6

The interesting take away there is that my custom hashmap ended up being only marginally faster than a normal `HashMap<String, _>` using fnv hashing.

**Comment by PSeitz** [comment]
Thanks for the heads up, yeah I am trying to create a fast string hashmap with round robin hashing. Since tantivy's hashmap is quite similar I copied it to learn from it (I'm not really familiar with implementing a hashmap) and benchmark it. I'm not sure yet combining tantivy's hashmap with round robin is the way to go, since the usage is a bit weird and the size always needs to be set upfront(?), but some concepts like referencing only the byte slice from a heap or vec instead of owning the string in the table seem reasonable.

> That's cool (as long as you don't forget to copy the license for the heap/expull stuff at one point), but the code you forked is actually "theoretically" bugged because I wrote it without caring at all about alignment.
In practise this has not been a problem so far, because x86 is fine and LLVM does not have any optimization relying on alignment there.

I just quickly copied some code together to test some scenarios and didn't notice the missing license in the stacker subfolder.
While skimming through I didn't spot alignment issues, is it related to BytesRef?

> The interesting take away there is that my custom hashmap ended up being only marginally faster than a normal HashMap<String, _> using fnv hashing.

Do you mean the tantivy's hashmap or the onehop-hashmap with custom hashmap? I also had some test runs where they are basically on par, but overall there is a ~15% speedup with this hashmap version mainly because of less hash calculations. Tantivy's hashmap can be 50% faster but also won't finish if split_memory is set too small.

**Comment by fulmicoton** [comment]
> The interesting take away there is that my custom hashmap ended up being only marginally faster than a normal HashMap<String, _> using fnv hashing.

`onehop-hashmap`. That's about the speedup I observed as well.
I was hoping having the key data and the value inlined one after the other would reduce the number of cache miss but I did not observe it, at least on the data size I worked on.

The new hashmap in tantivy (not in master yet) can handles resize.

**Comment by fulmicoton** [comment]
The alignment problem is kind of tricky. Basically I shouldn't do `&mut *ptr` when a pointer is not aligned.
I did not know that when I started developing tantivy.

**Comment by fulmicoton** [comment]
The specificity of our usage in search engine is :
- we want to accurately know the memory usage
- we want the memory usage to be as small as possible
- our keys have variable length, well it's strings
- we never remove stuff <---- that's an important one
- we expect the distribution of the number of key access to be very non-uniform among keys. For natural language, it follows a zipf distribution. As a result, a key inserted first is likely to be access more frequently than a key inserted later. (that does not mean cuckoo hash map are  a bad idea, but it is less beneficial than in the normal case.)

**Comment by PSeitz** [comment]
> The specificity of our usage in search engine is :
>
>    - we want to accurately know the memory usage
>    - we want the memory usage to be as small as possible
>    - our keys have variable length, well it's strings
>    - we never remove stuff <---- that's an important one
>    - we expect the distribution of the number of key access to be very non-uniform among keys. For natural language, it follows a zipf distribution. As a result, a key inserted first is likely to be access more frequently than a key inserted later. (that does not mean cuckoo hash map are a bad idea, but it is less beneficial than in the normal case.)


Yes, I agree. With a zipfs distribution it could be beneficial, to store early keys near each other in memory. Some fields although are likely to be unique keys. Another thing is that zipfs law possibly only applies to root words (https://en.wikipedia.org/wiki/Most_common_words_in_Spanish)

Plus points in our fst case would be a fast sorted key list extraction.

>    - we never remove stuff <---- that's an important one

Yes, depending on the hash table architecture, we don't need e.g. deleted flags, although resizing could involve some sort of deleting.

I switched the table now to use string references only, and store and reference them in an internal byte vec, similar to the heap, but opaque to the user. Performance is roughly on par with the current tantivy version in this repo. I didn't check the memory consumption yet, but the 24byte overhead per String is reduced to 8byte (start and length) currently.

I also have some rather crazy ideas, which I maybe give a try, like instead comparing the texts for identity, compare a second hash function, given it is unlikely enough for both hash functions to be equal for two strings. This could reduce memory io, because the actual string wouldn't be needed to be loaded in the insertion phase.

>I was hoping having the key data and the value inlined one after the other would reduce the number of cache miss but I did not observe it, at least on the data size I worked on.

Interesting try, I guess depending on the unsuccessful probe count, respectively the hashmaps load factor, where it just compares the hashed and not the strings this would lead to more cache misses 
with a higher load factor. Robin hood hashing could help here.

**Comment by PSeitz** [comment]
As an update, it's usable. String metadata is now 4byte for the position (so there's a 4Gb raw string limit) + the length vint encoded in the raw string vec. So it's possible to get a sorted string vec with 4 bytes per string (not yet implemented) and it's also possble to truncate the raw string vec, when building the hashmap is finished. I checked the memory consumption distribution, which is roughly for a 10Gb hashmap 4GB raw strings and 6GB hashes and key-values.

The structure of the table is currently pretty much the same, with two arrays for the hashes (4byte) and key-value pairs (key=stringref=4byte). Not sure if this is the best way, it's probably more compact when they are together.
